vivo-project/Vitro
# Moved password encryption from MD5 to a salted and secure hash - 1448

https://jira.duraspace.org/browse/VIVO-1448

**What does this pull request do?**

It replaces the MD5 password hashing with a more secure and salted password hashing algorithm.

**What's new?**

MD5 hashing has been replaced by Argon2i hashing, which is salted and configurable in terms of memory, time and parallelism needed to compute a hash. https://github.com/phxql/argon2-jvm

MD5 hashing has not been completely eliminated and the users with MD5 hash as their passwords are asked by the system to renew their password once they log in.

The configurable parameters of the Argon2i algorithm are stored in the "runtime.properties" file.

**How should this be tested?**

- Copy the useraccounts (with Md5 passwords) from an already installed instance of Vivo/Vitro and place them in the home directory of vitro with Argon2i hashing and check if the system promotes you to renew your password after logging in.
- Login using API with root credentials.  (Could be done using a browser i.e, http://[server address e.g., localhost:8080]/[application name e.g., vitro]/api/sparqlQuery?email=[root email]&password=[root password])
- Create a new user and login
- Change your password and login
- Create an invite/user/create-user link
- Check external authentical
- Change the values for parallelism, memory and time in runtime.properties and login ( you should be still able to login but the time required to login should change)
- Restrict user accounts and login
